src/sio/Sprint.3
.\"(c) Copyright 1992, 1993 by Panagiotis Tsirigotis
.\"All rights reserved.  The file named COPYRIGHT specifies the terms 
.\"and conditions for redistribution.
.\"
.\" $Id: Sprint.3,v 1.1 2001/05/26 22:04:50 seth Exp $
.TH Sprint 3X "29 May 1992"
.SH NAME
Sprint -- formatted stream output
.SH SYNOPSIS
.LP
.nf
.ft B
int Sprint( fd, format [ , ... ] )
int fd ;
char *format ;
.SH DESCRIPTION
\fBSprint()\fR provides formatted output conversion. The formatting is
controlled by the \fIformat\fR argument. All characters in
\fIformat\fR that do not specify a conversion are printed. A conversion
is specified by a '%' followed by a string that ends with a
conversion type character. The string may contain flags, a field width,
a precision, and a modifier.
.LP
Possible flags (more that one can be specified and they can be in any order)
include:
.TP 10
.B \'-'
specifies left adjustment of the converted argument. The default
is right adjustment. This flag is meaningful only if a field width
is specified.
.TP
.B \'+'
specifies that a number will always have a sign as a prefix (this
forces a '+' sign to appear if the number is positive).
.TP
.B \' '
prefixes a \fIspace\fR to the number if the number has not a sign
(therefore the '+' flag overrides this flag).
.TP
.B \'#'
The meaning of '#' depends on the conversion type character: for \fBo\fR 
conversions the first digit will be 0; for \fBx\fR or \fBX\fR conversions
\fB0x\fR or \fB0X\fR respectively will be prefixed to the number (if it
is not zero); for \fBe\fR, \fBE\fR, \fBf\fR, \fBg\fR, and \fBG\fR conversions
the number will always have a decimal point.
.TP
.B \'0'
specifies padding with zeros instead of spaces.
.LP
The field width is specified by a number. This number indicates the
\fIminimum\fR width for the field. A '*' may be used instead of the number.
In that case the width is the value of the next argument which should
be an \fBint\fR. 
A negative width value specifies left adjustment with a width equal
to the absolute width value.
.LP
A precision is specified by a '.' followed by a number. The meaning of
the precision depends on the type conversion character. For a string
conversion, precision determines how many characters are printed from
the string. For integer conversions, precision determines the 
number of digits used to print the number (zero padding is done if
the precision exceeds the length of the number). For floating point
conversions, precision determines the number of digits after the
decimal point (\fBe\fR, \fBE\fR, \fBf\fR) or the number of
significant digits (\fBg\fR, \fBG\fR).
A '*' may be used instead of a number to specify the precision. In that
case the precision is the value of the next argument which should
be an \fBint\fR.
The behavior of \fBSprint()\fR is undefined if the precision is negative.
.LP
The length modifier is \fBl\fR and indicates that the argument is
a \fBlong\fR integer.
.LP
The type conversion characters are:
\fBd, i, o, x, X, u, c, s, f, e, E, g, G, p, n, %\fR.
For floating point conversions the argument should be of type \fIdouble\fR.
.TP 10
.B d,i
specify signed decimal conversion.
.TP
.B u
specifies unsigned decimal conversion.
.TP
.B o
specifies octal conversion.
.TP
.B x,X
specify hexadecimal conversion. For 
.B x
the hex digits used are 0123456789abcdef. For
.B X
the hex digits used are 0123456789ABCDEF.
There is no leading
.B 0x
or
.B 0X
(use the '#' flag for that).
.TP
.B c
specifies character conversion; the argument should be of type
\fIchar\fR.
.TP
.B s
specifies string conversion; the argument should be of type
\fIchar *\fR.
.TP
.B f
specifies conversion to the form [-]ddd.dddd. The number
of digits after the decimal point depends on precision; the default is 6.
If the precision is 0, the decimal point is not printed (this can
be overriden with the '#' flag).
.B e,E
specify conversion to the form [-]ddd.dddd[eE][+-]ddd.
The number of digits after the decimal point depends on precision;
the default is 6. If the precision is 0, the decimal point is not printed
(this can be overriden with the '#' flag).
The exponent is at least 2 digit wide.
.TP
.B g,G
specify a conversion using the
.B e,E
format respectively if the
exponent is less than -4 or greater than or equal to the precision;
otherwise the 
.B f
format is used.
.TP
.B p
is used to print pointers (type \fIvoid *\fR,
or \fIchar *\fR if the compiler does not support the former).
.TP
.B n
expects a \fIint *\fR argument and puts in that integer 
the number of characters already printed by this call.
.TP
.B %
is used to print a \fI%\fR.
.LP
If an unknown conversion character is specified, the percent sign
followed by that character will be printed.
.SH RETURN VALUE
.LP
If no error occured, \fBSprint()\fR returns the number of characters
printed. In case of error, it returns \fBSIO_ERR\fR.
.SH BUGS
.LP
This is a list of differences between \fBSprint()\fR and the ANSI C Standard
\fBprintf()\fR:
.LP
\fBSprint()\fR does not support non-removal of trailing zeroes for
\fBg\fR and \fBG\fR conversions when the '#' flag is used.
.LP
\fBSprint()\fR does not support the h and L modifiers.
.LP
The current implementation assumes that \fIsizeof(int)==sizeof(long)\fR.
.LP
\fBSprint()\fR supports "%p" only if \fIsizeof(pointer)<=sizeof(int)\fR.


ChangeLog
2003-07-16  Seth Robertson  <seth@startide.sysdetect.com>

	* src/fsma/fsma.c (_fsm_free): Prevent infinite loops on fsm_free
	of corrupted memory

2003-07-09  James Tanis  <jtt@hawking.sysdetect.com>

	* src/dict/dll.c (dll_error_reason): Export pulibcally

2003-06-13  Alexander Dupuy  <dupuy@sysdetect.com>

	* include/dict.h, include/ht.h, include/pq.h: avoid stupid void * #1311

2003-06-07  Seth Robertson  <seth@startide.sysdetect.com>

	* src/dict/ht.c (ht_create): argsp may change size between
	external and internal users due to HASH_STATS define.

2003-05-16  Seth Robertson  <seth@startide.sysdetect.com>

	* src/dict/ht.c (ht_destroy): Allow clean up of all memory when
	insure is in use and fsma is replaced by malloc

2003-05-10  Alexander Dupuy  <dupuy@sysdetect.com>

	* include/dict.h (DICT_STATS): keep performance stats for hash tables

	* src/dict/htimpl.h (struct ht_stats): hash table performance stats
	(ENTRY_IS_FULL): kludge to detect bucket collisions

	* src/dict/ht.c: instrumented to keep performance stats for hash tables
	(print_hashstats): print performance stats

2003-04-28  Alexander Dupuy  <dupuy@sysdetect.com>

	* src/dict/ht.c (ht_delete): microscopic performance tweak
	[Seth got the big win already]

2003-04-16  James Tanis  <jtt@saint.sysdetect.com>

	* src/fsma/fsma.c (fsm_threaded_makeready): Only allow users to
	actually alter the value of thread_overried if we are compiled
	with threads.

2003-04-08  Seth Robertson  <seth@startide.sysdetect.com>

	* src/fsma/fsma.c (fsm_threaded_makeready): Provide interface for
	people to specify this will be a threaded program and to enforce
	thread safe memory protection.

2003-04-03  Seth Robertson  <seth@startide.sysdetect.com>

	* src/pq/hpq.c: Make thread-safe, through PQ_THREADED_SAFE

2003-03-31  Seth Robertson  <seth@winwood.sysdetect.com>

	* src/dict/*.c src/fsma/*.c: Make thread-safe, through
	DICT_THREADED_SAFE and DICT_THREADED_MEMORY

2002-12-10  James Tanis  <jtt@stevedore.sysdetect.com>

	* src/pq/Makefile: move -lpq to -lpqueue so as not to conflict
	with postgres

2002-11-11  James Tanis  <jtt@stevedore.sysdetect.com>

	* src/fsma/fsma.c, include/dict.h: opx_rollup_3 merge

2002-11-09  James Tanis  <jtt@stevedore.sysdetect.com>

	* include/dict.h, src/fsma/fsma.c: trheads merge

2002-11-06  Seth Robertson  <seth@stevedore.sysdetect.com>

	* src/fsma/fsma.c: Compile problems when using dmalloc

2002-11-05  Seth Robertson  <seth@stevedore.sysdetect.com>

	* include/dict.h: Define symbol for NOCOALESCE which makes it
	clear the only reason we are using it is because we are too lazy
	to make CLC thread-safe

2002-09-09  James Tanis  <jtt@stevedore.sysdetect.com>

	* src/dict/ht.c: insight nits

2002-09-05  Seth Robertson  <seth@stevedore.sysdetect.com>

	* src/dict/dll.c, ChangeLog: Do not create dlls w/o UNORDERED and
	w/o OOCMP function--coredump otherwise sometimes

2002-09-05  Seth Robertson  <seth@winwood.sysdetect.com>

	* src/dict/dll.c (dll_create): Default order is ordered.

	* src/dict/dll.c (dll_create): Use bitwise and for bits, instead
	of addition.  Prevented an argument sanity check from occurring.

2002-08-29  Seth Robertson  <seth@startide.sysdetect.com>

	* include/fsma.h (struct __fsma_header): AHHHHH-HHHHAAAAAAAA!
	The reference counter was previously a short.  Worker, for reasons
	too complicated to get into, can have more than 2^16th lists.
	If we happen (not unlikely) to destroy down to 2^16 (e.g. zero)
	fsma thinks all allocators are free and promptly proceeds to return
	the memory to the system.  Memory was free'd but still in use.
	BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA BAKA.

	* src/fsma/fsma.c (fsm_destroy): Perform assertion to verify we
	actually found the fsma allocator in the coalesce list.  Don't
	bother to realloc the coalesce list size down--it won't do no good
	nohow.

2002-08-27  Seth Robertson  <seth@startide.sysdetect.com>

	* src/dict/ht.c (ht_minimum): Serious and quite possibly
	bug-injecting misuse of a hint (cur_min).  The insert and delete
	routines were setting this to a bucket chain entry, not a table
	entry bucket (except when the insert routine had to perform a
	bucket chain expansion--when cur_min might be set to some other
	random value).  This usually was OK since bucket chain numbers are
	very low (0-3 range), but if a valid entry was in the table at
	slot zero or one, then ht_minimum might never see that item.

2002-08-14  James Tanis  <jtt@saint.sysdetect.com>

	* src/dict/ht.c (ht_minimum): Use cached minimum index as search start.
	(ht_do_insert): Update minimum index if less the current minimum.
	(ht_delete): Search for new minimum index if the index of the
	deleted node equals the current minimum.
	In addition add and obj_cnt so we don't try to perform expensive
	operations on empty hashes.

2002-05-13  James Tanis  <jtt@saint.sysdetect.com>

	* src/fsma/Makefile (BK_LARGE_LIB): Shared libfsma no longer
	breaks with libtool (on both linux and bsd).

2002-05-10  Alexander Dupuy  <dupuy@sysdetect.com>

	* src/fsma/Makefile (BK_LARGE_LIB): back out shared libfsma;
	this breaks libtool in a strange way that requires investigation

2002-05-10  James Tanis  <jtt@winwood.sysdetect.com>

	* src/fsma/fsma.c: Formatting and simplify coalesce code.

	* src/fsma/Makefile (BK_LARGE_LIB): Make libfsma shared.

	* src/dict/bst.c (bst_redelete): Improve efficiency by avoiding
	unnecessary checks and recusive calls.

2002-02-22  Alexander Dupuy  <dupuy@sysdetect.com>

	* include/dict.h (DICT_NUKE): delete contents and dict itself
	(_DICT_NUKE_WORK_LOOP): common code for DICT_NUKE, DICT_NUKE_CONTENTS

	* src/dict/dictimpl.h (HANDLE_ERROR): Eliminate unused id argument
	[use BK_FUNCNAME  here instead if you need it for debugging]

2002-01-31  Alexander Dupuy  <dupuy@sysdetect.com>

	* src/dict/Makefile (BK_LARGE_LIB): use ${LIBEXT} for libtool build

2002-01-09  Alexander Dupuy  <dupuy@sysdetect.com>

	* src/fsma/fsma.c: INSIGHT->INSURE


src/sio/sprint.c
/*
 * (c) Copyright 1992, 1993 by Panagiotis Tsirigotis
 * All rights reserved.  The file named COPYRIGHT specifies the terms
 * and conditions for redistribution.
 */



#include <ctype.h>

#include "sio.h"
#include "impl.h"

#ifdef _WIN32
#include <stdarg.h>
#endif /* _WIN32 */
#include "clchack.h"

#ifndef WIDE_INT
#define WIDE_INT		long
#endif

typedef WIDE_INT 		wide_int ;
typedef unsigned WIDE_INT 	u_wide_int ;
typedef int 			bool_int ;

#define S_NULL			"(null)"
#define S_NULL_LEN		6
#define FLOAT_DIGITS		6
#define EXPONENT_LENGTH		10

/*
 * NUM_BUF_SIZE is the size of the buffer used for arithmetic conversions
 *
 * XXX: this is a magic number; do not decrease it
 */
#define NUM_BUF_SIZE						512

PRIVATE char *conv_10( register wide_int, register bool_int, register bool_int *, char *, register int *);

/*
 * The INS_CHAR macro inserts a character in the buffer and writes
 * the buffer back to disk if necessary
 * It uses the char pointers sp and bep:
 * 	sp points to the next available character in the buffer
 * 	bep points to the end-of-buffer+1
 * While using this macro, note that the nextb pointer is NOT updated.
 *
 * No I/O is performed if fd is not positive. Negative fd values imply
 * conversion with the output directed to a string. Excess characters
 * are discarded if the string overflows.
 *
 * NOTE: Evaluation of the c argument should not have any side-effects
 */
#define INS_CHAR( c, sp, bep, odp, cc, fd )											\
			{																						\
				if ( sp < bep )																\
				{																					\
					*sp++ = c ;																	\
					cc++ ;																		\
				}																					\
				else																				\
				{																					\
					if ( fd >= 0 )																\
					{																				\
						odp->nextb = sp ;														\
						if ( __sio_writef( odp, fd ) != bep - odp->start )			\
							return( ( cc != 0 ) ? cc : SIO_ERR ) ;						\
						sp = odp->nextb ;														\
						*sp++ = c ;																\
						cc++ ;																	\
					}																				\
				}																					\
				if ( __SIO_MUST_FLUSH( *odp, c ) && fd >= 0 ) 						\
				{																					\
					int b_in_buffer = sp - odp->start ;									\
																									\
					odp->nextb = sp ;															\
					if ( __sio_writef( odp, fd ) != b_in_buffer )					\
						return( cc ) ;															\
					sp = odp->nextb ;															\
				}																					\
			}



#define NUM( c )			( c - '0' )

#define STR_TO_DEC( str, num )									\
									num = NUM( *str++ ) ;			\
									while ( isdigit( *str ) )		\
									{										\
										num *= 10 ;						\
										num += NUM( *str++ ) ;		\
									}

/*
 * This macro does zero padding so that the precision
 * requirement is satisfied. The padding is done by
 * adding '0's to the left of the string that is going
 * to be printed.
 */
#define FIX_PRECISION( adjust, precision, s, s_len )				\
					if ( adjust )												\
						while ( s_len < precision )						\
						{															\
							*--s = '0' ;										\
							s_len++ ;											\
						}

/*
 * Macro that does padding. The padding is done by printing
 * the character ch.
 */
#define PAD( width, len, ch )			do														\
												{														\
													INS_CHAR( ch, sp, bep, odp, cc, fd ) ;	\
													width-- ;										\
												}														\
												while ( width > len )

/*
 * Prefix the character ch to the string str
 * Increase length
 * Set the has_prefix flag
 */
#define PREFIX( str, length, ch )	*--str = ch ; length++ ; has_prefix = YES


/*
 * Sprint is the equivalent of printf for SIO.
 * It returns the # of chars written
 * Assumptions:
 *     - all floating point arguments are passed as doubles
 */
/* VARARGS2 */
int Sprint( int fd, char *fmt, ... )
{
	__sio_descriptor_t *dp = &__sio_descriptors[ fd ] ;
	register __sio_od_t *odp = ODP( dp ) ;
	register int cc ;
	va_list ap ;

	IO_SETUP( fd, dp, __SIO_OUTPUT_STREAM, SIO_ERR ) ;

	va_start( ap, fmt ) ;
	cc = __sio_converter( odp, fd, fmt, ap ) ;
	va_end( ap ) ;
	return( cc ) ;
}


/*
 * This is the equivalent of vfprintf for SIO
 */
int Sprintv( int fd, char *fmt, va_list ap)
{
	__sio_descriptor_t *dp = &__sio_descriptors[ fd ] ;
	register __sio_od_t *odp = ODP( dp ) ;

	IO_SETUP( fd, dp, __SIO_OUTPUT_STREAM, SIO_ERR ) ;
	return( __sio_converter( odp, fd, fmt, ap ) ) ;
}


/*
 * Convert a floating point number to a string formats 'f', 'e' or 'E'.
 * The result is placed in buf, and len denotes the length of the string
 * The sign is returned in the is_negative argument (and is not placed
 * in buf).
 */
PRIVATE char *conv_fp(register char format, register double num, boolean_e add_dp, int precision, bool_int *is_negative, char *buf, int *len)
				/* always add decimal point if YES */
{
	register char *s = buf ;
	register char *p ;
	int decimal_point ;
	char *ecvt(double, int, int *, int *), *fcvt(double, int, int *, int *) ;
	char *conv_10() ;
#ifndef __linux__				/* necessary kludge */
	char *strcpy(char *, const char *) ;
#endif

	if ( format == 'f' )
		p = fcvt( num, precision, &decimal_point, is_negative ) ;
	else /* either e or E format */
		p = ecvt( num, precision+1, &decimal_point, is_negative ) ;

	/*
	 * Check for Infinity and NaN
	 */
	if ( isalpha( *p ) )
	{
		*len = strlen( strcpy( buf, p ) ) ;
		*is_negative = FALSE ;
		return( buf ) ;
	}

	if ( format == 'f' )
		if ( decimal_point <= 0 )
		{
			*s++ = '0' ;
			if ( precision > 0 )
			{
				*s++ = '.' ;
				while ( decimal_point++ < 0 )
					*s++ = '0' ;
			}
			else if ( add_dp )
				*s++ = '.' ;
		}
		else
		{
			while ( decimal_point-- > 0 )
				*s++ = *p++ ;
			if ( precision > 0 || add_dp ) *s++ = '.' ;
		}
	else
	{
		*s++ = *p++ ;
		if ( precision > 0 || add_dp ) *s++ = '.' ;
	}

	/*
	 * copy the rest of p, the NUL is NOT copied
	 */
	while ( *p ) *s++ = *p++ ;		

	if ( format != 'f' )
	{
		char temp[ EXPONENT_LENGTH ] ;				/* for exponent conversion */
		int t_len ;
		bool_int exponent_is_negative ;

		*s++ = format ;		/* either e or E */
		decimal_point-- ;
		if ( decimal_point != 0 )
		{
			p = conv_10( (wide_int)decimal_point, FALSE, &exponent_is_negative,
												&temp[ EXPONENT_LENGTH ], &t_len ) ;
			*s++ = exponent_is_negative ? '-' : '+' ;
		
			/*
			 * Make sure the exponent has at least 2 digits
			 */
			if ( t_len == 1 )
				*s++ = '0' ;
			while ( t_len-- ) *s++ = *p++ ;
		}
		else
		{
			*s++ = '+' ;
			*s++ = '0' ;
			*s++ = '0' ;
		}
	}

	*len = s - buf ;
	return( buf ) ;
}


/*
 * Convert num to a base X number where X is a power of 2. nbits determines X.
 * For example, if nbits is 3, we do base 8 conversion
 * Return value:
 *			a pointer to a string containing the number
 *
 * The caller provides a buffer for the string: that is the buf_end argument
 * which is a pointer to the END of the buffer + 1 (i.e. if the buffer
 * is declared as buf[ 100 ], buf_end should be &buf[ 100 ])
 */
PRIVATE char *conv_p2(register u_wide_int num, register int nbits, char format, char *buf_end, register int *len)
{
	register int mask = ( 1 << nbits ) - 1 ;
	register char *p = buf_end ;
	static char low_digits[] = "0123456789abcdef" ;
	static char upper_digits[] = "0123456789ABCDEF" ;
	register char *digits = ( format == 'X' ) ? upper_digits : low_digits ;

	do
	{
		*--p = digits[ num & mask ] ;
		num >>= nbits ;
	}
	while( num ) ;

	*len = buf_end - p ;
	return( p ) ;
}



/*
 * Convert num to its decimal format.
 * Return value:
 *       - a pointer to a string containing the number (no sign)
 *			- len contains the length of the string
 *			- is_negative is set to TRUE or FALSE depending on the sign
 *			  of the number (always set to FALSE if is_unsigned is TRUE)
 *
 * The caller provides a buffer for the string: that is the buf_end argument
 * which is a pointer to the END of the buffer + 1 (i.e. if the buffer
 * is declared as buf[ 100 ], buf_end should be &buf[ 100 ])
 */
PRIVATE char *conv_10(register wide_int num, register bool_int is_unsigned, register bool_int *is_negative, char *buf_end, register int *len)
{
	register char *p = buf_end ;
	register u_wide_int magnitude ;

	if ( is_unsigned )
	{
		magnitude = (u_wide_int) num ;
		*is_negative = FALSE ;
	}
	else
	{
		*is_negative = ( num < 0 ) ;

		/*
		 * On a 2's complement machine, negating the most negative integer
		 * results in a number that cannot be represented as a signed integer.
		 * Here is what we do to obtain the number's magnitude:
		 *		a. add 1 to the number
		 *		b. negate it (becomes positive)
		 *		c. convert it to unsigned
		 *		d. add 1
		 */
		if ( *is_negative )
		{
			wide_int t = num + 1 ;

			magnitude = ( (u_wide_int) -t ) + 1 ;
		}
		else
			magnitude = (u_wide_int) num ;
	}

	/*
	 * We use a do-while loop so that we write at least 1 digit
	 */
	do
	{
		register u_wide_int new_magnitude = magnitude / 10 ;

		*--p = magnitude - new_magnitude*10 + '0' ;
		magnitude = new_magnitude ;
	}
	while ( magnitude ) ;

	*len = buf_end - p ;
	return( p ) ;
}


/*
 * Do format conversion placing the output in odp
 */
int __sio_converter(register __sio_od_t *odp, int fd, register char *fmt, va_list ap)
{
	register char *sp ;
	register char *bep ;
	register int cc = 0 ;
	register int i ;

	register char *s = NULL ;
	char *q ;
	int s_len ;

	register int min_width = 0 ;
	int precision = 0 ;
	enum { LEFT, RIGHT } adjust ;
	char pad_char ;
	char prefix_char ;

	double fp_num ;
	wide_int i_num = 0 ;
	u_wide_int ui_num ;

	char num_buf[ NUM_BUF_SIZE ] ;
	char char_buf[ 2 ] ;		/* for printing %% and %<unknown> */

	/*
	 * Flag variables
	 */
	boolean_e is_long ;
	boolean_e alternate_form ;
	boolean_e print_sign ;
	boolean_e print_blank ;
	boolean_e adjust_precision ;
	boolean_e adjust_width ;
	bool_int is_negative ;

	char *conv_10();
	char *gcvt(double, int, char *) ;
#ifndef __linux__				/* necessary kludge */
	char *strchr(const char *, int) ;
#endif

	sp = odp->nextb ;
	bep = odp->buf_end ;

	while ( *fmt )
	{
		if ( *fmt != '%' )
		{
			INS_CHAR( *fmt, sp, bep, odp, cc, fd ) ;
		}
		else
		{
			/*
			 * Default variable settings
			 */
			adjust = RIGHT ;
			alternate_form = print_sign = print_blank = NO ;
			pad_char = ' ' ;
			prefix_char = NUL ;

			fmt++ ;

			/*
			 * Try to avoid checking for flags, width or precision
			 */
			if ( isascii( *fmt ) && ! islower( *fmt ) )
			{
				/*
				 * Recognize flags: -, #, BLANK, +
				 */
				for ( ;; fmt++ )
				{
					if ( *fmt == '-' )
						adjust = LEFT ;
					else if ( *fmt == '+' )
						print_sign = YES ;
					else if ( *fmt == '#' )
						alternate_form = YES ;
					else if ( *fmt == ' ' )
						print_blank = YES ;
					else if ( *fmt == '0' )
						pad_char = '0' ;
					else
						break ;
				}

				/*
				 * Check if a width was specified
				 */
				if ( isdigit( *fmt ) )
				{
					STR_TO_DEC( fmt, min_width ) ;
					adjust_width = YES ;
				}
				else if ( *fmt == '*' )
				{
					min_width = va_arg( ap, int ) ;
					fmt++ ;
					adjust_width = YES ;
					if ( min_width < 0 )
					{
						adjust = LEFT ;
						min_width = -min_width ;
					}
				}
				else
					adjust_width = NO ;
			
				/*
				 * Check if a precision was specified
				 *
				 * XXX: an unreasonable amount of precision may be specified
				 *		  resulting in overflow of num_buf. Currently we
				 *		  ignore this possibility.
				 */
				if ( *fmt == '.' )
				{
					adjust_precision = YES ;
					fmt++ ;
					if ( isdigit( *fmt ) )
					{
						STR_TO_DEC( fmt, precision ) ;
					}
					else if ( *fmt == '*' )
					{
						precision = va_arg( ap, int ) ;
						fmt++ ;
						if ( precision < 0 )
							precision = 0 ;
					}
					else
						precision = 0 ;
				}
				else
					adjust_precision = NO ;
			}
			else
				adjust_precision = adjust_width = NO ;

			/*
			 * Modifier check
			 */
			if ( *fmt == 'l' )
			{
				is_long = YES ;
				fmt++ ;
			}
			else
				is_long = NO ;

			/*
			 * Argument extraction and printing.
			 * First we determine the argument type.
			 * Then, we convert the argument to a string.
			 * On exit from the switch, s points to the string that
			 * must be printed, s_len has the length of the string
			 * The precision requirements, if any, are reflected in s_len.
			 *
			 * NOTE: pad_char may be set to '0' because of the 0 flag.
			 *			It is reset to ' ' by non-numeric formats
			 */
			switch( *fmt )
			{
				case 'd':
				case 'i':
				case 'u':
					if ( is_long )
						i_num = va_arg( ap, wide_int ) ;
					else
						i_num = (wide_int) va_arg( ap, int ) ;
					s = conv_10( i_num, (*fmt) == 'u', &is_negative,
												&num_buf[ NUM_BUF_SIZE ], &s_len ) ;
					FIX_PRECISION( adjust_precision, precision, s, s_len ) ;

					if ( *fmt != 'u' )
					{
						if ( is_negative )
							prefix_char = '-' ;
						else if ( print_sign )
							prefix_char = '+' ;
						else if ( print_blank )
							prefix_char = ' ' ;
					}
					break ;
			

				case 'o':
					if ( is_long )
						ui_num = va_arg( ap, u_wide_int ) ;
					else
						ui_num = (u_wide_int) va_arg( ap, unsigned int ) ;
					s = conv_p2( ui_num, 3, *fmt,
												&num_buf[ NUM_BUF_SIZE ], &s_len ) ;
					FIX_PRECISION( adjust_precision, precision, s, s_len ) ;
					if ( alternate_form && *s != '0' )
					{
						*--s = '0' ;
						s_len++ ;
					}
					break ;
			

				case 'x':
				case 'X':
					if ( is_long )
						ui_num = (u_wide_int) va_arg( ap, u_wide_int ) ;
					else
						ui_num = (u_wide_int) va_arg( ap, unsigned int ) ;
					s = conv_p2( ui_num, 4, *fmt,
												&num_buf[ NUM_BUF_SIZE ], &s_len ) ;
					FIX_PRECISION( adjust_precision, precision, s, s_len ) ;
					if ( alternate_form && i_num != 0 )
					{
						*--s = *fmt ;			/* 'x' or 'X' */
						*--s = '0' ;
						s_len += 2 ;
					}
					break ;
			

				case 's':
					s = va_arg( ap, char * ) ;
					if ( s != NULL )
					{
						s_len = strlen( s ) ;
						if ( adjust_precision && precision < s_len )
							s_len = precision ;
					}
					else
					{
						s = S_NULL ;
						s_len = S_NULL_LEN ;
					}
					pad_char = ' ' ;
					break ;
			

				case 'f':
				case 'e':
				case 'E':
					fp_num = va_arg( ap, double ) ;

					s = conv_fp( *fmt, fp_num, alternate_form,
							( adjust_precision == NO ) ? FLOAT_DIGITS : precision,
													&is_negative, &num_buf[ 1 ], &s_len ) ;
					if ( is_negative )
						prefix_char = '-' ;
					else if ( print_sign )
						prefix_char = '+' ;
					else if ( print_blank )
						prefix_char = ' ' ;
					break ;


				case 'g':
				case 'G':
					if ( adjust_precision == NO )
						precision = FLOAT_DIGITS ;
					else if ( precision == 0 )
						precision = 1 ;
					/*
					 * We use &num_buf[ 1 ], so that we have room for the sign
					 */
					s = gcvt( va_arg( ap, double ), precision, &num_buf[ 1 ] ) ;
					if ( *s == '-' )
						prefix_char = *s++ ;
					else if ( print_sign )
						prefix_char = '+' ;
					else if ( print_blank )
						prefix_char = ' ' ;

					s_len = strlen( s ) ;

					if ( alternate_form && ( q = strchr( s, '.' ) ) == NULL )
						s[ s_len++ ] = '.' ;
					if ( *fmt == 'G' && ( q = strchr( s, 'e' ) ) != NULL )
						*q = 'E' ;
					break ;


				case 'c':
					char_buf[ 0 ] = (char) (va_arg( ap, int )) ;
					s = &char_buf[ 0 ] ;
					s_len = 1 ;
					pad_char = ' ' ;
					break ;


				case '%':
					char_buf[ 0 ] = '%' ;
					s = &char_buf[ 0 ] ;
					s_len = 1 ;
					pad_char = ' ' ;
					break ;


				case 'n':
					*(va_arg( ap, int * )) = cc ;
					break ;

				/*
				 * Always extract the argument as a "char *" pointer. We
				 * should be using "void *" but there are still machines
				 * that don't understand it.
				 * If the pointer size is equal to the size of an unsigned
				 * integer we convert the pointer to a hex number, otherwise
				 * we print "%p" to indicate that we don't handle "%p".
				 */
				case 'p':
					ui_num = (u_wide_int) va_arg( ap, char * ) ;

					if ( sizeof( char * ) <= sizeof( u_wide_int ) )
						s = conv_p2( ui_num, 4, 'x',
												&num_buf[ NUM_BUF_SIZE ], &s_len ) ;
					else
					{
						s = "%p" ;
						s_len = 2 ;
					}
					pad_char = ' ' ;
					break ;
				

				case NUL:
					/*
					 * The last character of the format string was %.
					 * We ignore it.
					 */
					continue ;


					/*
				 	 * The default case is for unrecognized %'s.
					 * We print %<char> to help the user identify what
					 * option is not understood.
					 * This is also useful in case the user wants to pass
					 * the output of __sio_converter to another function
					 * that understands some other %<char> (like syslog).
					 * Note that we can't point s inside fmt because the
					 * unknown <char> could be preceded by width etc.
					 */
				default:
					char_buf[ 0 ] = '%' ;
					char_buf[ 1 ] = *fmt ;
					s = char_buf ;
					s_len = 2 ;
					pad_char = ' ' ;
					break ;
			}

			if ( prefix_char != NUL )
			{
				*--s = prefix_char ;
				s_len++ ;
			}

			if ( adjust_width && adjust == RIGHT && min_width > s_len )
			{
				if ( pad_char == '0' && prefix_char != NUL )
				{
					INS_CHAR( *s, sp, bep, odp, cc, fd )
					s++ ;
					s_len-- ;
					min_width-- ;
				}
				PAD( min_width, s_len, pad_char ) ;
			}

			/*
			 * Print the string s.
			 */
			for ( i = s_len ; i != 0 ; i-- )
			{
				INS_CHAR( *s, sp, bep, odp, cc, fd ) ;
				s++ ;
			}

			if ( adjust_width && adjust == LEFT && min_width > s_len )
				PAD( min_width, s_len, pad_char ) ;
		}
		fmt++ ;
	}
	odp->nextb = sp ;
	return( cc ) ;
}

